src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,122 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPASITY = 10;
+    private static final int NOT_FOUND = -1;
+    private int capasity;
+    private int size;
+    private T[] array;
+
+    public ArrayList() {
+        capasity = DEFAULT_CAPASITY;
+        array = (T[]) new Object[DEFAULT_CAPASITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == capasity) {
+            grow();
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (invalidIndexToAdd(index)) {
+            throw new ArrayListIndexOutOfBoundsException(index + " is out of range index");
+        }
+        if (size == capasity) {
+            grow();
+        }
+        for (int i = size; i > index; i--) {
+            array[i] = array[i - 1];
+        }
+        array[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (invalidIndexToSetGet(index)) {
+            throw new ArrayListIndexOutOfBoundsException(index + " is out of range index");
+        }
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (invalidIndexToSetGet(index)) {
+            throw new ArrayListIndexOutOfBoundsException(index + " is out of range index");
+        }
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T value = get(index);
+        for (int i = index; i < size - 1; i++) {
+            array[i] = array[i + 1];
+        }
+        size--;
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = find(element);
+        if (index == NOT_FOUND) {
+            throw new NoSuchElementException("Element " + element + " was not found");
+        }
+        return remove(index);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void grow() {
+        int oldCapasity = capasity;
+        capasity += capasity / 2;
```

Comment:
better create a clearly defined and readable GROW_FACTOR constant

Reply:
\+